metaflow/plugins/argo/event_decorators.py

```diff
@@ -0,0 +1,247 @@
+import re
+
+from metaflow import current
+from metaflow.decorators import FlowDecorator, StepDecorator
+from metaflow.exception import MetaflowException
+from .util import are_events_configured, valid_statuses
+
+ERR_MSG = """Make sure configuration entries METAFLOW_EVENT_SOURCE_NAME, METAFLOW_EVENT_SOURCE_URL, and 
+METAFLOW_EVENT_SERVICE_ACCOUNT are correct.
+
+Authentication (Optional)
+=========================
+If authentication is required and credentials are stored in a Kubernetes secret, check
+METAFLOW_EVENT_AUTH_SECRET and METAFLOW_EVENT_AUTH_KEY.
+
+If an authentication token is used, verify METAFLOW_EVENT_AUTH_TOKEN has the correct value.
+"""
+
+BAD_EVENT_ERR_MSG = """"The event field is empty. Verify the event name begins with an alphanumeric
+character and doesn't begin with the reserved prefix 'metaflow_'.
+"""
+
+
+class BadEventNameException(MetaflowException):
+    headline = "Bad or missing event name detected"
+
+
+class BadEventMappingsException(MetaflowException):
+    headline = "Bad event field mappings detected"
+
+
+def sanitize_user_event(name):
+    # NATS uses dots as a subtopic delimiter
+    name = name.replace(".", "-")
+    name = name.lower()
+    # METAFLOW_ and metaflow_ prefixes are reserved
+    return re.sub("^metaflow_", "", name)
+
+
+def validate_mappings(mappings):
+    if mappings is None:
+        return None
+    if type(mappings) != dict:
+        raise BadEventMappingsException(
+            msg="Event field mappings must be a dict of string keys and "
+            + "values mapping event fields to flow parameters."
+        )
+    for k in mappings.keys():
+        v = mappings[k]
+        if type(k) != str or type(v) != str:
+            raise BadEventMappingsException(
+                msg="Event field mappings must be a dict of string keys and "
+                + "values mapping event fields to flow parameters."
+            )
+    return mappings
+
+
+def validate_data(deco_name, data):
+    for key in data.keys():
+        if type(key) != str:
+            raise MetaflowException(
+                msg="@%s requires data to be a dict with string keys: %s"
+                % (deco_name, str(key))
+            )
+    return data
+
+
+class TriggerOnDecorator(FlowDecorator):
```

Comment:
Can we decouple this to explicit `TriggerOnFinish` and `Trigger` decorators? That way, we can have a proper delineation that `TriggerOnFinish` is a first-class metaflow concept that underlying workflow orchestrators may not support out of the box. For example - Airflow doesn't have any notion of events, but someday we would like to introduce `TriggerOnFinish` for Airflow (native support for Airflow events is unlikely since they don't exist at the moment).